examples/terraform-provider-primes/provider.go

```diff
@@ -0,0 +1,144 @@
+					{
+						Name:              "number",
+						Type:              tftypes.Number,
+						RequiredForImport: true,
+					},
+				},
+			},
+		},
+	}, nil
+}
+
+func (p PrimeNumberProvider) GetProviderSchema(ctx context.Context, request *tfprotov5.GetProviderSchemaRequest) (*tfprotov5.GetProviderSchemaResponse, error) {
+	return &tfprotov5.GetProviderSchemaResponse{
+		Provider: &tfprotov5.Schema{
+			Block: &tfprotov5.SchemaBlock{
+				Description:     "Prime Provider",
+				DescriptionKind: tfprotov5.StringKindPlain,
+			},
+		},
+		ResourceSchemas: map[string]*tfprotov5.Schema{
+			"prime": p.primeSchema(),
+		},
+		ListResourceSchemas: map[string]*tfprotov5.Schema{
+			"prime": p.primeSchema(),
+		},
+	}, nil
+}
+
+func (p PrimeNumberProvider) primeSchema() *tfprotov5.Schema {
+	return &tfprotov5.Schema{
+		Block: &tfprotov5.SchemaBlock{
+			Attributes: []*tfprotov5.SchemaAttribute{
+				{
+					Name:            "number",
+					Type:            tftypes.Number,
+					Description:     "The nth prime",
+					DescriptionKind: tfprotov5.StringKindPlain,
+					Optional:        true,
+				},
+				{
+					Name:            "ordinal",
+					Type:            tftypes.Number,
+					Description:     "n",
+					DescriptionKind: tfprotov5.StringKindPlain,
+					Optional:        true,
+				},
+			},
+		},
+	}
+}
+
+func (p PrimeNumberProvider) PrepareProviderConfig(ctx context.Context, request *tfprotov5.PrepareProviderConfigRequest) (*tfprotov5.PrepareProviderConfigResponse, error) {
+	return &tfprotov5.PrepareProviderConfigResponse{
+		PreparedConfig: request.Config,
+	}, nil
+}
+
+func (p PrimeNumberProvider) ConfigureProvider(ctx context.Context, request *tfprotov5.ConfigureProviderRequest) (*tfprotov5.ConfigureProviderResponse, error) {
+	return &tfprotov5.ConfigureProviderResponse{}, nil
+}
+
+func (p PrimeNumberProvider) convertToDynamicValue(number int, ordinal int) (tfprotov5.DynamicValue, error) {
+	typ := p.primeSchema().ValueType()
+	value := map[string]tftypes.Value{
+		"number":  tftypes.NewValue(tftypes.Number, number),
+		"ordinal": tftypes.NewValue(tftypes.Number, ordinal),
+	}
+
+	return tfprotov5.NewDynamicValue(typ, tftypes.NewValue(typ, value))
+}
+
+func (p PrimeNumberProvider) ListResource(ctx context.Context, request *tfprotov5.ListResourceRequest) (*tfprotov5.ListResourceServerStream, error) {
+	primes := []int{2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97}
+
+	results := func(push func(tfprotov5.ListResourceResult) bool) {
+		for i, prime := range primes {
+			ordinal := i + 1
+			displayName := fmt.Sprintf("primes[%d]: %d", ordinal, prime)
+
+			resourceObject, err := p.convertToDynamicValue(prime, ordinal)
+			if err != nil {
+				panic(err)
+			}
+
+			protoEv := tfprotov5.ListResourceResult{
+				DisplayName: displayName,
+				Resource:    &resourceObject,
+			}
+			if !push(protoEv) {
+				fmt.Println("let's stop here")
```

Comment:
curious: why do we stop? Is this the case where the caller effectively cancels?

Reply:
>  Is this the case where the caller effectively cancels?

Yes! I think I wrote a code comment (in Framework?) that documents this and I will port it over here.